### PR TITLE
Scripts to generate RPMs for must-gather-clean

### PR DESCRIPTION
**Pull request by arjunrn**
Added scripts and make target to cross-compile the must-gather-clean binary and create RPMs.

### CODE REVIEW

File: hack/compile.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+fi
+
+for platform in linux-amd64 linux-arm64 linux-ppc64le linux-s390x darwin-amd64 windows-amd64 ; do
+  echo "Cross compiling $platform and placing binary at dist/bin/$platform/"
+  if [ $platform == "windows-amd64" ]; then
+    GOARCH=amd64 GOOS=windows go build -o dist/bin/$platform/must-gather-clean.exe "${@}" ./cmd/must-gather-clean/
+  else
+    GOARCH=${platform#*-} GOOS=${platform%-*} go build -o dist/bin/$platform/must-gather-clean "${@}" ./cmd/must-gather-clean/
+  fi
+done
```
Comment (tjungblu) [inline hack/compile.sh:20]: nit: newline

File: hack/prepare-release.sh [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    echo "generating SHA256_SUM for release packages"
+    release_dir_files=$(find $RELEASE_DIR -maxdepth 1 ! -name SHA256_SUM -type f -printf "%f\n")
+    for filename in $release_dir_files; do
+        sha_sum=$(sha256sum $RELEASE_DIR/"${filename}"|awk '{ print $1 }'); echo "$sha_sum"  "$filename";
+    done > ${RELEASE_DIR}/SHA256_SUM
+    echo "The SHA256 SUM for the release packages are:"
+    cat ${RELEASE_DIR}/SHA256_SUM
+}
+
+release_sha
```
Comment (tjungblu) [inline hack/prepare-release.sh:47]: also nit newline

File: hack/rpm-prepare.sh [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+# we want to do a re-lease of same version as we are not backporting.
+export MGC_VERSION=${MGC_VERSION:=0.0.1}
+export MGC_RELEASE=${MGC_RELEASE:=1}
+
+export GIT_COMMIT=${GIT_COMMIT:=$(git rev-parse --short HEAD 2>/dev/null)}
+export MGC_RPM_VERSION=${MGC_VERSION//-}
+
+# Golang version variables, if you are bumping this, please contact redhat maintainers to ensure that internal
+# build systems can handle these versions
+export GOLANG_VERSION=${GOLANG_VERSION:-1.15}
```
Comment (tjungblu) [inline hack/rpm-prepare.sh:16]: in gomod we run 1.16, I also don't see you setting that value anywhere. Any reason for 1.15 as the default?

File: rpms/must-gather-clean.spec [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+URL:            https://github.com/openshift/must-gather-clean/tree/%{mgc_cli_version}
+
+Source0:        %{source_tar}
+BuildRequires:  gcc
+BuildRequires:  golang >= %{golang_version}
+Provides:       %{package_name} = %{mgc_version}
+Obsoletes:      %{package_name} <= %{mgc_version}
+
+%description
+must-gather-clean is a fast, iterative, and straightforward CLI tool for developers who write, build, and deploy applications on OpenShift.
```
Comment (tjungblu) [inline rpms/must-gather-clean.spec:33]: is it? :)

File: cmd/must-gather-clean/version.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	VERSION = "v0.0.1"
```
Comment (tjungblu) [inline cmd/must-gather-clean/version.go:10]: can you add a script for bumping this too? and possibly also add this into prepare-release.

File: hack/rpm-local-build.sh [hunk 6]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+set -euo pipefail
+
+if [[ ! -d dist/rpmbuild  ]]; then
+	echo "Cannot build as artifacts are not generated. Run scrips/rpm-prepare.sh first"
```
Comment (tjungblu) [inline hack/rpm-local-build.sh:5]: hack/rpm-prepare.sh

File: hack/rpm-prepare.sh (hunk 3 above)
Comment (tjungblu) [reply hack/rpm-prepare.sh:16]: matter of fact, I do get a pretty weird error when I run this with 1.16:

```
[tjungblu ~/git/arjmg]$ hack/rpm-local-build.sh 
Building locally
error: Failed build dependencies:
	golang >= 1.15 is needed by must-gather-clean-0.0.1-1.el8.x86_64
[tjungblu ~/git/arjmg]$ go version
go version go1.16.4 linux/amd64

```

File: Makefile [hunk 7]
```diff
@@ -1,3 +1,8 @@
+COMMON_GOFLAGS := -mod=vendor
```
Comment (tjungblu) [inline Makefile:1]: could you also add some documentation in the DEV readme?

File: hack/rpm-local-build.sh [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+	exit 1
```
Comment (pweil-) [inline hack/rpm-local-build.sh:6]: any reason not to just run it?

File: hack/rpm-prepare.sh [hunk 9]
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+echo "Reading MGC_VERSION, MCG_RELEASE and GIT_COMMIT env, if they are set"
+# Change version as needed. In most cases MGC_RELEASE would not be touched unless
+# we want to do a re-lease of same version as we are not backporting.
+export MGC_VERSION=${MGC_VERSION:=0.0.1}
```
Comment (pweil-) [inline hack/rpm-prepare.sh:8]: if these are unset should this default to something that makes it obviously a non-production version like `dirty-{gitCommitRev}`?

File: hack/rpm-prepare.sh [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+echo "Generating spec file $SPEC_DIR/must-gather-clean.spec"
+envsubst <rpms/must-gather-clean.spec > $SPEC_DIR/must-gather-clean.spec
+
+echo "Generating tarball $SOURCES_DIR/$NAME.tar.gz"
+# Copy code for manipulation
+cp -arf ./* $SOURCES_DIR/$NAME
+pushd $SOURCES_DIR
+pushd $NAME
+# Remove bin if it exists, we dont need it in tarball
+rm -rf ./odo
```
Comment (pweil-) [inline hack/rpm-prepare.sh:58]: Why is `odo` present? Seemed a bit random and more of an artifact of a dirty build environment.

Comment (arjunrn) [reply hack/rpm-prepare.sh:58]: Close, these config files and shell scripts were copied over from another project called `openshift-do` or `odo`.

File: hack/rpm-prepare.sh (hunk 3 above)
Comment (arjunrn) [reply hack/rpm-prepare.sh:16]: Copy-pasta :)

File: Makefile (hunk 7 above)
Comment (arjunrn) [reply Makefile:1]: I think it's better to add the documentation once the release pipeline is in place.

File: cmd/must-gather-clean/version.go (hunk 5 above)
Comment (arjunrn) [reply cmd/must-gather-clean/version.go:10]: Added the bump script. The `prepare-release.sh` script doesn't really deal with version information. So I don't want to bump the versions in there. But if that's going to be single entrypoint for building releases I will add it later.

File: hack/rpm-local-build.sh (hunk 8 above)
Comment (arjunrn) [reply hack/rpm-local-build.sh:6]: This script is taken from [openshift/odo](https://github.com/openshift/odo/blob/main/scripts/rpm-local-build.sh). Right now I'm keeping the script as similar to their as possible. If this  can be optimized I will do that after we have generated our first release.

File: hack/rpm-prepare.sh (hunk 9 above)
Comment (arjunrn) [reply hack/rpm-prepare.sh:8]: I would like that as well but the release pipeline depends on manually updating the release version number.

File: rpms/must-gather-clean.spec [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+%license LICENSE
+%dir %{_datadir}/%{name}-redistributable
+%{_datadir}/%{name}-redistributable/must-gather-clean-linux-amd64
+%{_datadir}/%{name}-redistributable/must-gather-clean-linux-amd64.tar.gz
+%{_datadir}/%{name}-redistributable/must-gather-clean-linux-arm64
+%{_datadir}/%{name}-redistributable/must-gather-clean-linux-arm64.tar.gz
+%{_datadir}/%{name}-redistributable/must-gather-clean-linux-ppc64le
+%{_datadir}/%{name}-redistributable/must-gather-clean-linux-ppc64le.tar.gz
+%{_datadir}/%{name}-redistributable/must-gather-clean-linux-s390x
+%{_datadir}/%{name}-redistributable/must-gather-clean-linux-s390x.tar.gz
```
Comment (tjungblu) [inline rpms/must-gather-clean.spec:93]: could you also remove the ppc and s3 builds please? just talked with the odo manager and they mentioned that was just a specific IBM release that they also needed to specifically support and they got assistance when building against it.

Comment (arjunrn) [reply rpms/must-gather-clean.spec]: Removed.

**Review by tjungblu — APPROVED** [review]
/lgtm 

let's get this in, we should get our dist-git repo by today as well

**Pull request merged**